File: packages/typescript/server/src/merkle.ts
/**
 * RFC 6962 Merkle tree with inclusion and consistency proofs.
 *
 * Port of the Python implementation — produces identical roots for the same
 * input data (protocol interop requirement).
 */
import { createHash } from "crypto";

function sha256(data: Buffer): Buffer {
  return createHash("sha256").update(data).digest();
}

function leafHash(data: Buffer): Buffer {
  return sha256(Buffer.concat([Buffer.from([0x00]), data]));
}

function nodeHash(left: Buffer, right: Buffer): Buffer {
  return sha256(Buffer.concat([Buffer.from([0x01]), left, right]));
}

function hex(digest: Buffer): string {
  return `sha256:${digest.toString("hex")}`;
}

function unhex(s: string): Buffer {
  if (s.startsWith("sha256:")) {
    return Buffer.from(s.slice(7), "hex");
  }
  return Buffer.from(s, "hex");
}

function largestPowerOf2LessThan(n: number): number {
  if (n <= 1) return 0;
  let k = 1;
  while (k * 2 < n) k *= 2;
  return k;
}

export interface InclusionStep {
  hash: string;
  side: "left" | "right";
}

export interface Snapshot {
  root: string;
  leaf_count: number;
}

export class MerkleTree {
  private _leaves: Buffer[] = [];

  get leafCount(): number {
    return this._leaves.length;
  }

  get root(): string {
    if (this._leaves.length === 0) {
      return hex(sha256(Buffer.alloc(0)));
    }
    return hex(this._computeRoot(0, this._leaves.length));
  }

  addLeaf(data: Buffer): void {
    this._leaves.push(leafHash(data));
  }

  snapshot(): Snapshot {
    return { root: this.root, leaf_count: this.leafCount };
  }

  // --- Inclusion proofs ---

  inclusionProof(index: number): InclusionStep[] {
    if (index < 0 || index >= this._leaves.length) {
      throw new RangeError(
        `Leaf index ${index} out of range [0, ${this._leaves.length})`
      );
    }
    const path: InclusionStep[] = [];
    this._buildInclusionPath(index, 0, this._leaves.length, path);
    return path;
  }

  verifyInclusion(
    index: number,
    data: Buffer,
    proof: InclusionStep[],
    expectedRoot?: string
  ): boolean {
    return MerkleTree.verifyInclusionStatic(
      data,
      proof,
      expectedRoot ?? this.root
    );
  }

  static verifyInclusionStatic(
    data: Buffer,
    proof: InclusionStep[],
    expectedRoot: string
  ): boolean {
    let current = leafHash(data);
    for (const step of proof) {
      const sibling = Buffer.from(step.hash, "hex");
      if (step.side === "left") {
        current = nodeHash(sibling, current);
      } else {
        current = nodeHash(current, sibling);
      }
    }
    return hex(current) === expectedRoot;
  }

  // --- Consistency proofs ---

  consistencyProof(oldSize: number): Buffer[] {
    if (oldSize < 0 || oldSize > this._leaves.length) {
      throw new RangeError(`oldSize ${oldSize} out of range`);
    }
    if (oldSize === 0 || oldSize === this._leaves.length) {
      return [];
    }
    return this._subproof(oldSize, 0, this._leaves.length, true);
  }

  verifyConsistency(
    oldRoot: string,
    oldSize: number,
    proof: Buffer[]
  ): boolean {
    return MerkleTree.verifyConsistencyStatic(
      oldRoot,
      oldSize,
      this.root,
      this.leafCount,
      proof
    );
  }

  static verifyConsistencyStatic(
    oldRoot: string,
    oldSize: number,
    newRoot: string,
    newSize: number,
    proof: Buffer[]
  ): boolean {
    if (oldSize === 0) return true;
    if (oldSize === newSize) return oldRoot === newRoot && proof.length === 0;
    if (proof.length === 0) return false;
    try {
      const [oldHash, newHash] = verifyConsistency(
        oldSize,
        newSize,
        proof,
        unhex(oldRoot)
      );
      return hex(oldHash) === oldRoot && hex(newHash) === newRoot;
    } catch {
      return false;
    }
  }

  // --- Internal tree computation ---

  private _computeRoot(lo: number, hi: number): Buffer {
    const n = hi - lo;
    if (n === 1) return this._leaves[lo];
    const split = largestPowerOf2LessThan(n);
    const left = this._computeRoot(lo, lo + split);
    const right = this._computeRoot(lo + split, hi);
    return nodeHash(left, right);
  }

  private _buildInclusionPath(
    index: number,
    lo: number,
    hi: number,
    path: InclusionStep[]
  ): void {
    const n = hi - lo;
    if (n === 1) return;
    const split = largestPowerOf2LessThan(n);
    if (index - lo < split) {
      this._buildInclusionPath(index, lo, lo + split, path);
      const right = this._computeRoot(lo + split, hi);
      path.push({ hash: right.toString("hex"), side: "right" });
    } else {
      this._buildInclusionPath(index, lo + split, hi, path);
      const left = this._computeRoot(lo, lo + split);
      path.push({ hash: left.toString("hex"), side: "left" });
    }
  }

  private _subproof(
    m: number,
    lo: number,
    hi: number,
    start: boolean
  ): Buffer[] {
    const n = hi - lo;
    if (m === n) {
      if (!start) return [this._computeRoot(lo, hi)];
      return [];
    }
    if (m === 0) {
      return [this._computeRoot(lo, hi)];
    }
    const k = largestPowerOf2LessThan(n);
    if (m <= k) {
      return [
        ...this._subproof(m, lo, lo + k, start),
        this._computeRoot(lo + k, hi),
      ];
    } else {
      return [
        ...this._subproof(m - k, lo + k, hi, false),
        this._computeRoot(lo, lo + k),
      ];
    }
  }
}

// --- Consistency proof verification (standalone) ---

function verifyConsistency(
  oldSize: number,
  newSize: number,
  proof: Buffer[],
  oldRootBytes: Buffer
): [Buffer, Buffer] {
  let idx = 0;

  function consume(): Buffer {
    if (idx >= proof.length) throw new RangeError("proof too short");
    return proof[idx++];
  }

  function walk(
    m: number,
    n: number,
    start: boolean
  ): [Buffer, Buffer] {
    if (m === n) {
      if (!start) {
        const p = consume();
        return [p, p];
      }
      return [oldRootBytes, oldRootBytes];
    }
    if (m === 0) {
      const p = consume();
      return [null as unknown as Buffer, p];
    }
    const k = largestPowerOf2LessThan(n);
    if (m <= k) {
      const [oldLeft, newLeft] = walk(m, k, start);
      const rightHash = consume();
      const newCombined = nodeHash(newLeft, rightHash);
      return [oldLeft, newCombined];
    } else {
      const [oldRight, newRight] = walk(m - k, n - k, false);
      const leftHash = consume();
      const newCombined = nodeHash(leftHash, newRight);
      const oldCombined = nodeHash(leftHash, oldRight);
      return [oldCombined, newCombined];
    }
  }

  const [oldHash, newHash] = walk(oldSize, newSize, true);
  if (idx !== proof.length) {
    throw new Error(`proof has extra elements: used ${idx} of ${proof.length}`);
  }
  return [oldHash, newHash];
}

File: examples/anip/anip_flight_demo/primitives/merkle.py
"""RFC 6962 Merkle tree with inclusion and consistency proofs."""
import hashlib
from typing import Any


def _sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def _leaf_hash(data: bytes) -> bytes:
    """RFC 6962 §2.1: MTH({d(0)}) = SHA-256(0x00 || d(0))"""
    return _sha256(b"\x00" + data)


def _node_hash(left: bytes, right: bytes) -> bytes:
    """RFC 6962 §2.1: SHA-256(0x01 || left || right)"""
    return _sha256(b"\x01" + left + right)


def _hex(digest: bytes) -> str:
    return f"sha256:{digest.hex()}"


def _unhex(s: str) -> bytes:
    """Inverse of _hex: 'sha256:<hex>' -> bytes."""
    if s.startswith("sha256:"):
        return bytes.fromhex(s[7:])
    return bytes.fromhex(s)


def _is_power_of_2(n: int) -> bool:
    return n > 0 and (n & (n - 1)) == 0


def _largest_power_of_2_less_than(n: int) -> int:
    if n <= 1:
        return 0
    k = 1
    while k * 2 < n:
        k *= 2
    return k


class MerkleTree:
    """Append-only Merkle tree following RFC 6962 structure."""

    def __init__(self) -> None:
        self._leaves: list[bytes] = []

    @property
    def leaf_count(self) -> int:
        return len(self._leaves)

    @property
    def root(self) -> str:
        if not self._leaves:
            return _hex(_sha256(b""))
        return _hex(self._compute_root(0, len(self._leaves)))

    def add_leaf(self, data: bytes) -> None:
        self._leaves.append(_leaf_hash(data))

    def snapshot(self) -> dict[str, Any]:
        return {"root": self.root, "leaf_count": self.leaf_count}

    # --- Inclusion proofs ---

    def inclusion_proof(self, index: int) -> list[dict[str, Any]]:
        if index < 0 or index >= len(self._leaves):
            raise IndexError(f"Leaf index {index} out of range [0, {len(self._leaves)})")
        path: list[dict[str, Any]] = []
        self._build_inclusion_path(index, 0, len(self._leaves), path)
        return path

    def verify_inclusion(self, index: int, data: bytes, proof: list[dict[str, Any]],
                         expected_root: str | None = None) -> bool:
        return self.verify_inclusion_static(data, proof, expected_root or self.root)

    @staticmethod
    def verify_inclusion_static(data: bytes, proof: list[dict[str, Any]],
                                expected_root: str) -> bool:
        current = _leaf_hash(data)
        for step in proof:
            sibling = bytes.fromhex(step["hash"])
            if step["side"] == "left":
                current = _node_hash(sibling, current)
            else:
                current = _node_hash(current, sibling)
        return _hex(current) == expected_root

    # --- Consistency proofs ---

    def consistency_proof(self, old_size: int) -> list[bytes]:
        if old_size < 0 or old_size > len(self._leaves):
            raise ValueError(f"old_size {old_size} out of range")
        if old_size == 0 or old_size == len(self._leaves):
            return []
        return self._subproof(old_size, 0, len(self._leaves), True)

    def _subproof(self, m: int, lo: int, hi: int, start: bool) -> list[bytes]:
        """RFC 6962 §2.1.4 SUBPROOF(m, D[lo:hi], start)."""
        n = hi - lo
        if m == n:
            if not start:
                return [self._compute_root(lo, hi)]
            return []
        if m == 0:
            return [self._compute_root(lo, hi)]
        k = _largest_power_of_2_less_than(n)
        if m <= k:
            return (self._subproof(m, lo, lo + k, start)
                    + [self._compute_root(lo + k, hi)])
        else:
            return (self._subproof(m - k, lo + k, hi, False)
                    + [self._compute_root(lo, lo + k)])

    @staticmethod
    def verify_consistency_static(
        old_root: str, old_size: int, new_root: str, new_size: int,
        proof: list[bytes]
    ) -> bool:
        """Verify a consistency proof between two tree sizes.

        Uses the verification algorithm from RFC 6962 §2.1.4, matching
        the SUBPROOF generation structure.
        """
        if old_size == 0:
            return True
        if old_size == new_size:
            return old_root == new_root and len(proof) == 0
        if not proof:
            return False
        try:
            old_hash, new_hash = _verify_consistency(
                old_size, new_size, proof, _unhex(old_root)
            )
        except (IndexError, ValueError):
            return False
        return _hex(old_hash) == old_root and _hex(new_hash) == new_root

    # --- Internal tree computation ---

    def _compute_root(self, lo: int, hi: int) -> bytes:
        n = hi - lo
        if n == 1:
            return self._leaves[lo]
        split = _largest_power_of_2_less_than(n)
        left = self._compute_root(lo, lo + split)
        right = self._compute_root(lo + split, hi)
        return _node_hash(left, right)

    def _build_inclusion_path(
        self, index: int, lo: int, hi: int, path: list[dict[str, Any]]
    ) -> None:
        n = hi - lo
        if n == 1:
            return
        split = _largest_power_of_2_less_than(n)
        if index - lo < split:
            self._build_inclusion_path(index, lo, lo + split, path)
            right = self._compute_root(lo + split, hi)
            path.append({"hash": right.hex(), "side": "right"})
        else:
            self._build_inclusion_path(index, lo + split, hi, path)
            left = self._compute_root(lo, lo + split)
            path.append({"hash": left.hex(), "side": "left"})


def _verify_consistency(
    old_size: int, new_size: int, proof: list[bytes], old_root_bytes: bytes
) -> tuple[bytes, bytes]:
    """Verify consistency proof by replaying the SUBPROOF structure.

    The proof is generated by SUBPROOF(m, D[0:n], b=True). We replay the
    same recursive decomposition, consuming proof elements in order, and
    reconstruct both old_hash and new_hash.

    When the subproof hits m==n with start=True (emitting no element),
    the hash is the old tree root which the verifier already knows.
    """
    idx_box = [0]

    def consume() -> bytes:
        i = idx_box[0]
        if i >= len(proof):
            raise IndexError("proof too short")
        idx_box[0] = i + 1
        return proof[i]

    def walk(m: int, n: int, start: bool) -> tuple[bytes, bytes]:
        """Returns (old_hash_contribution, new_hash_contribution).

        For subtrees entirely within the old tree (m==n, start=True),
        old_hash == new_hash == the known old subtree root. Since the
        subproof emits nothing here, we use old_root_bytes which the
        verifier supplies. This works because when start=True and m==n,
        it only occurs once at the deepest left recursion, and that
        subtree IS the old tree root.
        """
        if m == n:
            if not start:
                p = consume()
                return (p, p)
            # start=True, m==n: this subtree is exactly the old tree.
            # No proof element. Use the claimed old root.
            return (old_root_bytes, old_root_bytes)
        if m == 0:
            # This subtree is entirely new; doesn't affect old hash.
            p = consume()
            return (None, p)  # type: ignore[return-value]
        k = _largest_power_of_2_less_than(n)
        if m <= k:
            old_left, new_left = walk(m, k, start)
            right_hash = consume()
            new_combined = _node_hash(new_left, right_hash)
            return (old_left, new_combined)
        else:
            old_right, new_right = walk(m - k, n - k, False)
            left_hash = consume()
            new_combined = _node_hash(left_hash, new_right)
            old_combined = _node_hash(left_hash, old_right)
            return (old_combined, new_combined)

    old_hash, new_hash = walk(old_size, new_size, True)
    if idx_box[0] != len(proof):
        raise ValueError(f"proof has extra elements: used {idx_box[0]} of {len(proof)}")
    return old_hash, new_hash
